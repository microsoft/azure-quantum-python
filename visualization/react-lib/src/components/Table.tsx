import React from "react";

interface Props {
  rowGroups: RowGroup[];
}

export interface RowGroup {
  header: string;
  rows: Row[];
}

export interface Row {
  data: string[][];
}

const Table: React.FC<Props> = ({ rowGroups }) => {
  return (
    <table>
      {rowGroups.map((rowGroup, groupIndex) => (
        <tbody key={groupIndex}>
          <tr>
            <th style={{ textAlign: "left" }}>{rowGroup.header}</th>
          </tr>
          {rowGroup.rows.map((row, rowIndex) => (
            <tr key={rowIndex}>
              {row.data.map((cell, cellIndex) => (
                <>
                  <td
                    key={`${rowIndex}-${cellIndex}`}
                    style={{ paddingLeft: "10px" }}
                  >
                    {cell[0]}:
                  </td>
                  <td
                    key={`${rowIndex}-${cellIndex + 1}`}
                    style={{ paddingLeft: "10px" }}
                  >
                    <span style={{ fontWeight: "bold", textAlign: "right" }}>
                      {cell[1]}
                    </span>
                  </td>
                </>
              ))}
            </tr>
          ))}
        </tbody>
      ))}
    </table>
  );
};

export default Table;
