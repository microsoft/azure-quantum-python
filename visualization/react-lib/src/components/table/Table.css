.tooltip {
  padding: 5px;
  text-align: center;
  align-items: center;
  animation: tooltipFadeIn ease-in 0.2s;
  border-radius: 6px;
  &::after {
    content: "";
    position: absolute;
    width: 0;
    height: 0;
    border-left: 5px solid transparent;
    border-right: 5px solid transparent;
    border-top: 5px solid #cccccc;
    top: -5px;
    left: 50%;
    transform: translateX(-50%);
  }
}

@keyframes tooltipFadeIn {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}

.table-element{
  justify-content: center;
}
