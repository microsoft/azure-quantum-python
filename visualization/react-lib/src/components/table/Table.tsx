/*------------------------------------
  Copyright (c) Microsoft Corporation.
  Licensed under the MIT License.
  All rights reserved.
------------------------------------ */
import * as React from "react";
import {
  DetailsList,
  IColumn,
  IDetailsGroupRenderProps,
  IDetailsList,
  IGroup,
  SelectionMode,
} from "@fluentui/react";
import { getTheme, mergeStyleSets } from "@fluentui/react/lib/Styling";

const ROW_HEIGHT = 42; // from DEFAULT_ROW_HEIGHTS in DetailsRow.styles.ts
const GROUP_HEADER_AND_FOOTER_SPACING = 8;
const GROUP_HEADER_AND_FOOTER_BORDER_WIDTH = 1;
const GROUP_HEADER_HEIGHT = 95;
const GROUP_FOOTER_HEIGHT: number =
  GROUP_HEADER_AND_FOOTER_SPACING * 4 +
  GROUP_HEADER_AND_FOOTER_BORDER_WIDTH * 2;

const theme = getTheme();

const classNames = mergeStyleSets({
  headerAndFooter: {
    borderTop: `${GROUP_HEADER_AND_FOOTER_BORDER_WIDTH}px solid ${theme.palette.neutralQuaternary}`,
    borderBottom: `${GROUP_HEADER_AND_FOOTER_BORDER_WIDTH}px solid ${theme.palette.neutralQuaternary}`,
    padding: GROUP_HEADER_AND_FOOTER_SPACING,
    margin: `${GROUP_HEADER_AND_FOOTER_SPACING}px 0`,
    background: theme.palette.neutralLighterAlt,
    // Overlay the sizer bars
    position: "relative",
    zIndex: 100,
  },
  headerTitle: [
    theme.fonts.large,
    {
      padding: "4px 0",
    },
  ],
});

export interface IItem {
  name: string;
  value: string;
  description: string;
}

export interface IState {
  items: IItem[];
  groups: IGroup[];
  showItemIndexInView: boolean;
  isCompactMode: boolean;
}

export class TableComponent extends React.Component<
  { state: IState; columns: IColumn[] },
  IState
> {
  private _root = React.createRef<IDetailsList>();
  private _columns: IColumn[];

  constructor(props: { state: IState; columns: IColumn[] }) {
    super(props);

    this.state = props.state;
    this._columns = props.columns;
  }

  public componentWillUnmount() {
    if (this.state.showItemIndexInView) {
      const itemIndexInView = this._root.current!.getStartItemIndexInView();
      alert("first item index that was in view: " + itemIndexInView);
    }
  }

  public render() {
    const { items, groups } = this.state;

    return (
      <div>
        <DetailsList
          componentRef={this._root}
          items={items}
          groups={groups}
          columns={this._columns}
          groupProps={{
            onRenderHeader: this._onRenderGroupHeader,
            showEmptyGroups: false,
          }}
          compact={true}
          indentWidth={4}
          getGroupHeight={this._getGroupHeight}
          selectionMode={SelectionMode.none}
          isHeaderVisible={false}
        />
      </div>
    );
  }

  private _onRenderGroupHeader: IDetailsGroupRenderProps["onRenderHeader"] = (
    props,
  ) => {
    if (props) {
      return (
        <div className={classNames.headerAndFooter}>
          <div className={classNames.headerTitle}>{`${props.group!.name}`}</div>
        </div>
      );
    }

    return null;
  };

  private _getGroupTotalRowHeight = (group: IGroup): number => {
    return group.isCollapsed ? 0 : ROW_HEIGHT * group.count;
  };

  private _getGroupHeight = (group: IGroup): number => {
    return (
      GROUP_HEADER_HEIGHT +
      GROUP_FOOTER_HEIGHT +
      this._getGroupTotalRowHeight(group)
    );
  };
}
