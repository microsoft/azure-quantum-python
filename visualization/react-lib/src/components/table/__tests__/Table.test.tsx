/*------------------------------------
  Copyright (c) Microsoft Corporation.
  Licensed under the MIT License.
  All rights reserved.
------------------------------------ */
import React from "react";
import { create } from "react-test-renderer";
import { IColumn, IGroup, ThemeProvider } from "@fluentui/react";
import { Icon } from "@fluentui/react/lib/Icon";
import {
  getTheme,
  mergeStyleSets,
  setIconOptions,
} from "@fluentui/react/lib/Styling";
import { TooltipHost } from "@fluentui/react/lib/Tooltip";

import { IItem, IState, TableComponent } from "../Table";
// Suppress icon warnings.
setIconOptions({
  disableWarnings: true,
});

const classNames = mergeStyleSets({
  cellText: {
    overflow: "hidden",
    textOverflow: "ellipsis",
  },
  tooltipHost: {
    marginLeft: "8px",
    cursor: "default",
  },
  infoIcon: {
    width: "12px",
    height: "12px",
    display: "inline-block",
    verticalAlign: "-0.1rem",
    fill: getTheme().semanticColors.infoIcon,
  },
});

describe("Table tests", () => {
  it("Verify Table", () => {
    const tableItems: IItem[] = [
      {
        name: "Total physical qubits",
        value: "12",
        description:
          "Total physical qubits required for algorithm and T factories.",
      },
      {
        name: "Physical T factory qubits",
        value: "20",
        description: "Number of physical qubits for the T factories.",
      },
      {
        name: "Number of T factory copies",
        value: "100",
        description:
          "Number of T factories capable of producing the demanded T states during the algorithm's runtime.",
      },
      {
        name: "Physical qubits for single T factory",
        value: "2",
        description: "",
      },
    ];

    const tableGroups: IGroup[] = [
      {
        key: "1",
        name: "Group 1",
        startIndex: 0,
        count: 1,
      },
      {
        key: "2",
        name: "Group 2",
        startIndex: 1,
        count: 2,
      },
      {
        key: "3",
        name: "Group 3",
        startIndex: 3,
        count: 1,
      },
    ];

    const tableProps: IState = {
      items: tableItems,
      groups: tableGroups,
      showItemIndexInView: false,
      isCompactMode: false,
    };

    const columns: IColumn[] = [
      {
        key: "name",
        name: "Name",
        onRender: (item: IItem) => {
          return (
            <div className={classNames.cellText} data-is-focusable={true}>
              {item.name}
              {item.description ? (
                <TooltipHost
                  hostClassName={classNames.tooltipHost}
                  content={item.description}
                >
                  <Icon iconName="Info" className={classNames.infoIcon} />
                </TooltipHost>
              ) : (
                <></>
              )}
            </div>
          );
        },
        minWidth: 220,
        flexGrow: 3,
      },
      {
        key: "value",
        name: "Value",
        onRender: (item: IItem) => {
          return (
            <div className={classNames.cellText} data-is-focusable={true}>
              {item.value}
            </div>
          );
        },
        minWidth: 50,
        flexGrow: 1,
      },
    ];

    const component = create(
      <ThemeProvider>
        <TableComponent state={tableProps} columns={columns} />
      </ThemeProvider>,
    );

    expect(component.toJSON()).toMatchSnapshot("Table");
  });
});
