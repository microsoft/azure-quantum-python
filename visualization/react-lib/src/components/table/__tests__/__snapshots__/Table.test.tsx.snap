// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table tests Verify Table: Table 1`] = `
<div
  className="body-114"
>
  <div>
    <div
      className="ms-Viewport"
      style={
        {
          "minHeight": 1,
          "minWidth": 1,
        }
      }
    >
      <div
        className="ms-DetailsList ms-DetailsList--Compact is-horizontalConstrained root-115"
        data-automationid="DetailsList"
        data-is-scrollable="false"
      >
        <div
          aria-colcount={3}
          aria-rowcount={7}
          role="grid"
        >
          <div
            className="ms-DetailsList-headerWrapper"
            onKeyDown={[Function]}
            role="presentation"
          />
          <div
            className="ms-DetailsList-contentWrapper"
            onKeyDown={[Function]}
            role="presentation"
          >
            <div
              className="ms-SelectionZone"
              onClick={[Function]}
              onContextMenu={[Function]}
              onDoubleClick={[Function]}
              onFocusCapture={[Function]}
              onKeyDown={[Function]}
              onKeyDownCapture={[Function]}
              onMouseDown={[Function]}
              onMouseDownCapture={[Function]}
              role="presentation"
            >
              <div
                className="ms-FocusZone css-120 ms-GroupedList ms-GroupedList--Compact root-117 focusZone-116"
                data-automationid="GroupedList"
                data-focuszone-id="FocusZone3"
                data-is-scrollable="false"
                onBlur={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onMouseDownCapture={[Function]}
                role="presentation"
              >
                <div
                  className="ms-List"
                  role="presentation"
                >
                  <div
                    className="ms-List-surface"
                    role="presentation"
                  >
                    <div
                      className="ms-List-page"
                      role="presentation"
                      style={{}}
                    >
                      <div
                        className="ms-List-cell"
                        data-automationid="ListCell"
                        data-list-index={0}
                        role="presentation"
                      >
                        <div
                          className="ms-GroupedList-group group-118"
                          role="presentation"
                        >
                          <div
                            className="headerAndFooter-109"
                          >
                            <div
                              className="headerTitle-110"
                            >
                              Group 1
                            </div>
                          </div>
                          <div
                            aria-label="Group 1"
                            className="ms-List"
                            id="GroupedListSection4"
                            role="rowgroup"
                          >
                            <div
                              className="ms-List-surface"
                              role="presentation"
                            >
                              <div
                                className="ms-List-page"
                                role="presentation"
                                style={{}}
                              >
                                <div
                                  className="ms-List-cell"
                                  data-automationid="ListCell"
                                  data-list-index={0}
                                  role="presentation"
                                >
                                  <div
                                    aria-rowindex={2}
                                    className="ms-FocusZone css-120 ms-DetailsRow root-123"
                                    data-automationid="DetailsRow"
                                    data-focuszone-id="FocusZone8"
                                    data-is-focusable={true}
                                    data-item-index={0}
                                    data-selection-index={0}
                                    data-selection-touch-invoke={true}
                                    id="row2-0"
                                    onFocus={[Function]}
                                    onKeyDown={[Function]}
                                    onMouseDownCapture={[Function]}
                                    role="row"
                                    tabIndex={0}
                                  >
                                    <span
                                      className="ms-GroupSpacer"
                                      role="gridcell"
                                      style={
                                        {
                                          "display": "inline-block",
                                          "width": 4,
                                        }
                                      }
                                    />
                                    <div
                                      className="ms-DetailsRow-fields fields-129"
                                      data-automationid="DetailsRowFields"
                                      role="presentation"
                                    >
                                      <div
                                        className="ms-DetailsRow-cell cell-126 cellUnpadded-124"
                                        data-automation-key="name"
                                        data-automationid="DetailsRowCell"
                                        role="gridcell"
                                        style={
                                          {
                                            "width": 240,
                                          }
                                        }
                                      >
                                        <div
                                          className="cellText-111"
                                          data-is-focusable={true}
                                        >
                                          Total physical qubits
                                          <div
                                            className="ms-TooltipHost tooltipHost-132"
                                            onBlurCapture={[Function]}
                                            onFocusCapture={[Function]}
                                            onKeyDown={[Function]}
                                            onMouseEnter={[Function]}
                                            onMouseLeave={[Function]}
                                            role="none"
                                          >
                                            <i
                                              aria-hidden={true}
                                              className="infoIcon-133"
                                              data-icon-name="Info"
                                            />
                                            <div
                                              hidden={true}
                                              id="tooltip9"
                                              style={
                                                {
                                                  "border": 0,
                                                  "height": 1,
                                                  "margin": -1,
                                                  "overflow": "hidden",
                                                  "padding": 0,
                                                  "position": "absolute",
                                                  "whiteSpace": "nowrap",
                                                  "width": 1,
                                                }
                                              }
                                            >
                                              Total physical qubits required for algorithm and T factories.
                                            </div>
                                          </div>
                                        </div>
                                      </div>
                                      <div
                                        className="ms-DetailsRow-cell cell-126 cellUnpadded-124"
                                        data-automation-key="value"
                                        data-automationid="DetailsRowCell"
                                        role="gridcell"
                                        style={
                                          {
                                            "width": 70,
                                          }
                                        }
                                      >
                                        <div
                                          className="cellText-111"
                                          data-is-focusable={true}
                                        >
                                          12
                                        </div>
                                      </div>
                                    </div>
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      className="ms-List-page"
                      role="presentation"
                      style={{}}
                    >
                      <div
                        className="ms-List-cell"
                        data-automationid="ListCell"
                        data-list-index={1}
                        role="presentation"
                      >
                        <div
                          className="ms-GroupedList-group group-118"
                          role="presentation"
                        >
                          <div
                            className="headerAndFooter-109"
                          >
                            <div
                              className="headerTitle-110"
                            >
                              Group 2
                            </div>
                          </div>
                          <div
                            aria-label="Group 2"
                            className="ms-List"
                            id="GroupedListSection5"
                            role="rowgroup"
                          >
                            <div
                              className="ms-List-surface"
                              role="presentation"
                            >
                              <div
                                className="ms-List-page"
                                role="presentation"
                                style={{}}
                              >
                                <div
                                  className="ms-List-cell"
                                  data-automationid="ListCell"
                                  data-list-index={1}
                                  role="presentation"
                                >
                                  <div
                                    aria-rowindex={4}
                                    className="ms-FocusZone css-120 ms-DetailsRow root-123"
                                    data-automationid="DetailsRow"
                                    data-focuszone-id="FocusZone11"
                                    data-is-focusable={true}
                                    data-item-index={1}
                                    data-selection-index={1}
                                    data-selection-touch-invoke={true}
                                    id="row2-1"
                                    onFocus={[Function]}
                                    onKeyDown={[Function]}
                                    onMouseDownCapture={[Function]}
                                    role="row"
                                  >
                                    <span
                                      className="ms-GroupSpacer"
                                      role="gridcell"
                                      style={
                                        {
                                          "display": "inline-block",
                                          "width": 4,
                                        }
                                      }
                                    />
                                    <div
                                      className="ms-DetailsRow-fields fields-129"
                                      data-automationid="DetailsRowFields"
                                      role="presentation"
                                    >
                                      <div
                                        className="ms-DetailsRow-cell cell-126 cellUnpadded-124"
                                        data-automation-key="name"
                                        data-automationid="DetailsRowCell"
                                        role="gridcell"
                                        style={
                                          {
                                            "width": 240,
                                          }
                                        }
                                      >
                                        <div
                                          className="cellText-111"
                                          data-is-focusable={true}
                                        >
                                          Physical T factory qubits
                                          <div
                                            className="ms-TooltipHost tooltipHost-132"
                                            onBlurCapture={[Function]}
                                            onFocusCapture={[Function]}
                                            onKeyDown={[Function]}
                                            onMouseEnter={[Function]}
                                            onMouseLeave={[Function]}
                                            role="none"
                                          >
                                            <i
                                              aria-hidden={true}
                                              className="infoIcon-133"
                                              data-icon-name="Info"
                                            />
                                            <div
                                              hidden={true}
                                              id="tooltip12"
                                              style={
                                                {
                                                  "border": 0,
                                                  "height": 1,
                                                  "margin": -1,
                                                  "overflow": "hidden",
                                                  "padding": 0,
                                                  "position": "absolute",
                                                  "whiteSpace": "nowrap",
                                                  "width": 1,
                                                }
                                              }
                                            >
                                              Number of physical qubits for the T factories.
                                            </div>
                                          </div>
                                        </div>
                                      </div>
                                      <div
                                        className="ms-DetailsRow-cell cell-126 cellUnpadded-124"
                                        data-automation-key="value"
                                        data-automationid="DetailsRowCell"
                                        role="gridcell"
                                        style={
                                          {
                                            "width": 70,
                                          }
                                        }
                                      >
                                        <div
                                          className="cellText-111"
                                          data-is-focusable={true}
                                        >
                                          20
                                        </div>
                                      </div>
                                    </div>
                                  </div>
                                </div>
                                <div
                                  className="ms-List-cell"
                                  data-automationid="ListCell"
                                  data-list-index={2}
                                  role="presentation"
                                >
                                  <div
                                    aria-rowindex={5}
                                    className="ms-FocusZone css-120 ms-DetailsRow root-123"
                                    data-automationid="DetailsRow"
                                    data-focuszone-id="FocusZone14"
                                    data-is-focusable={true}
                                    data-item-index={2}
                                    data-selection-index={2}
                                    data-selection-touch-invoke={true}
                                    id="row2-2"
                                    onFocus={[Function]}
                                    onKeyDown={[Function]}
                                    onMouseDownCapture={[Function]}
                                    role="row"
                                  >
                                    <span
                                      className="ms-GroupSpacer"
                                      role="gridcell"
                                      style={
                                        {
                                          "display": "inline-block",
                                          "width": 4,
                                        }
                                      }
                                    />
                                    <div
                                      className="ms-DetailsRow-fields fields-129"
                                      data-automationid="DetailsRowFields"
                                      role="presentation"
                                    >
                                      <div
                                        className="ms-DetailsRow-cell cell-126 cellUnpadded-124"
                                        data-automation-key="name"
                                        data-automationid="DetailsRowCell"
                                        role="gridcell"
                                        style={
                                          {
                                            "width": 240,
                                          }
                                        }
                                      >
                                        <div
                                          className="cellText-111"
                                          data-is-focusable={true}
                                        >
                                          Number of T factory copies
                                          <div
                                            className="ms-TooltipHost tooltipHost-132"
                                            onBlurCapture={[Function]}
                                            onFocusCapture={[Function]}
                                            onKeyDown={[Function]}
                                            onMouseEnter={[Function]}
                                            onMouseLeave={[Function]}
                                            role="none"
                                          >
                                            <i
                                              aria-hidden={true}
                                              className="infoIcon-133"
                                              data-icon-name="Info"
                                            />
                                            <div
                                              hidden={true}
                                              id="tooltip15"
                                              style={
                                                {
                                                  "border": 0,
                                                  "height": 1,
                                                  "margin": -1,
                                                  "overflow": "hidden",
                                                  "padding": 0,
                                                  "position": "absolute",
                                                  "whiteSpace": "nowrap",
                                                  "width": 1,
                                                }
                                              }
                                            >
                                              Number of T factories capable of producing the demanded T states during the algorithm's runtime.
                                            </div>
                                          </div>
                                        </div>
                                      </div>
                                      <div
                                        className="ms-DetailsRow-cell cell-126 cellUnpadded-124"
                                        data-automation-key="value"
                                        data-automationid="DetailsRowCell"
                                        role="gridcell"
                                        style={
                                          {
                                            "width": 70,
                                          }
                                        }
                                      >
                                        <div
                                          className="cellText-111"
                                          data-is-focusable={true}
                                        >
                                          100
                                        </div>
                                      </div>
                                    </div>
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      className="ms-List-page"
                      role="presentation"
                      style={{}}
                    >
                      <div
                        className="ms-List-cell"
                        data-automationid="ListCell"
                        data-list-index={2}
                        role="presentation"
                      >
                        <div
                          className="ms-GroupedList-group group-118"
                          role="presentation"
                        >
                          <div
                            className="headerAndFooter-109"
                          >
                            <div
                              className="headerTitle-110"
                            >
                              Group 3
                            </div>
                          </div>
                          <div
                            aria-label="Group 3"
                            className="ms-List"
                            id="GroupedListSection6"
                            role="rowgroup"
                          >
                            <div
                              className="ms-List-surface"
                              role="presentation"
                            >
                              <div
                                className="ms-List-page"
                                role="presentation"
                                style={{}}
                              >
                                <div
                                  className="ms-List-cell"
                                  data-automationid="ListCell"
                                  data-list-index={3}
                                  role="presentation"
                                >
                                  <div
                                    aria-rowindex={7}
                                    className="ms-FocusZone css-120 ms-DetailsRow root-123"
                                    data-automationid="DetailsRow"
                                    data-focuszone-id="FocusZone17"
                                    data-is-focusable={true}
                                    data-item-index={3}
                                    data-selection-index={3}
                                    data-selection-touch-invoke={true}
                                    id="row2-3"
                                    onFocus={[Function]}
                                    onKeyDown={[Function]}
                                    onMouseDownCapture={[Function]}
                                    role="row"
                                  >
                                    <span
                                      className="ms-GroupSpacer"
                                      role="gridcell"
                                      style={
                                        {
                                          "display": "inline-block",
                                          "width": 4,
                                        }
                                      }
                                    />
                                    <div
                                      className="ms-DetailsRow-fields fields-129"
                                      data-automationid="DetailsRowFields"
                                      role="presentation"
                                    >
                                      <div
                                        className="ms-DetailsRow-cell cell-126 cellUnpadded-124"
                                        data-automation-key="name"
                                        data-automationid="DetailsRowCell"
                                        role="gridcell"
                                        style={
                                          {
                                            "width": 240,
                                          }
                                        }
                                      >
                                        <div
                                          className="cellText-111"
                                          data-is-focusable={true}
                                        >
                                          Physical qubits for single T factory
                                        </div>
                                      </div>
                                      <div
                                        className="ms-DetailsRow-cell cell-126 cellUnpadded-124"
                                        data-automation-key="value"
                                        data-automationid="DetailsRowCell"
                                        role="gridcell"
                                        style={
                                          {
                                            "width": 70,
                                          }
                                        }
                                      >
                                        <div
                                          className="cellText-111"
                                          data-is-focusable={true}
                                        >
                                          2
                                        </div>
                                      </div>
                                    </div>
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
